packages/loopring_v3/contracts/amm/libamm/AmmJoinProcess.sol

```diff
@@ -83,109 +37,178 @@ library AmmJoinProcess
         )
         internal
     {
-        S.validatePoolTransaction(
+        S.checkPoolTxApproval(
             join.owner,
-            AmmJoinRequest.hashPoolJoin(ctx.domainSeparator, join),
+            AmmJoinRequest.hash(ctx.domainSeparator, join),
             signature
         );
 
+        bool joinFromLayer1 = (
+            join.direction == AmmData.Direction.L1_TO_L1 ||
+            join.direction == AmmData.Direction.L1_TO_L2
+        );
+
+        if (joinFromLayer1) {
+            require(signature.length == 0, "LAYER1_JOIN_WITH_OFFCHAIN_APPROVAL_DISALLOWED");
+
+            if (join.nonce == S.joinLocks[msg.sender].length) {
+                S.joinLocks[msg.sender].pop();
+            } else {
+                delete S.joinLocks[msg.sender][join.nonce - 1];
+
+                if (S.joinLockIdx[msg.sender] == join.nonce - 1) {
+                    S.joinLockIdx[msg.sender]++;
+                }
+            }
+        } else {
+            require(join.nonce == 0, "LAYER2_JOIN_WITH_NONCE_DISALLOWED");
+        }
+
         // Check if the requirements are fulfilled
-        (bool slippageRequirementMet, uint poolAmountOut, uint96[] memory amounts) = _calculateJoinAmounts(S, ctx, join);
+        (bool slippageOK, uint96 mintAmount, uint96[] memory amounts) = _calculateJoinAmounts(ctx, join);
 
-        if (!slippageRequirementMet) return;
+        if (!slippageOK) {
+            if (joinFromLayer1) {
```

Comment:
If `joinAmounts[i] != amounts[i]` we don't refund and leave the funds of the user locked inside this contract? (locked up forever until perhaps the pool is shutdown and then divided by the remaining LPs). But if that is the goal (not sure if that's a good idea) we may as well deposit the full amount to the pool when joining from L1.

Reply:
fixed